scanner/src/main/kotlin/ScanResultsCache.kt

```diff
@@ -75,17 +77,44 @@ interface ScanResultsCache {
 
         var stats = CacheStatistics()
 
-        fun configure(config: ArtifactoryCacheConfiguration) {
-            require(config.url.isNotBlank()) {
-                "URL for Artifactory cache is missing."
+        fun configure(config: CacheConfiguration) {
+            config.validate()
+
+            if (config is ArtifactoryCacheConfiguration) {
+                cache = ArtifactoryCache(config)
+                log.info { "Using Artifactory cache '${config.url}'." }
+            } else if (config is CloudStorageCacheConfiguration) {
+                cache = CloudStorageCache(config)
+                log.info { "Using Cloud Storage cache." }
+            }
+        }
+
+        override fun add(id: Identifier, scanResult: ScanResult): Boolean = run {
+            // Do not cache empty scan results. It is likely that something went wrong when they were created, and if not,
+            // it is cheap to re-create them.
+            if (scanResult.summary.fileCount == 0) {
+                log.info { "Not caching scan result for '$id' because no files were scanned." }
+
+                return false
             }
 
-            require(config.apiToken.isNotBlank()) {
-                "API token for Artifactory cache is missing."
+            // Do not cache scan results without raw result. The raw result can be set to null for other usages, but in the
+            // cache it must never be null.
+            if (scanResult.rawResult == null) {
+                log.info { "Not caching scan result for '$id' because the raw result is null." }
+
+                return false
+            }
+
+            // Do not cache scan results without provenance information, because they cannot be assigned to the revision of
+            // the package source code later.
+            if (scanResult.provenance.sourceArtifact == null && scanResult.provenance.vcsInfo == null) {
+                log.info { "Not caching scan result for '$id' because no provenance information is available." }
+
+                return false
             }
 
-            cache = ArtifactoryCache(config.url, config.apiToken)
-            log.info { "Using Artifactory cache '${config.url}'." }
+            cache.add(id, scanResult)
```

Comment:
Would have been nice to have this refactoring in a separate commit, because refactorings should not be mixed with new features in the same commit.

Reply:
@mnonnenmacher Unless you're feeling strongly about this one, I'd leave as is since it would require some effort to  work backwards and move it to separate commit — but noted for further contributions, I'll try to be more diligent with my commits (I'm used to squashing everything once it's merges into master :D)

Reply:
Yes, please keep it as is and make smaller commits in future.